PLATUI-1877: sbt-a11y-linter - productionise the AutomaticAccessibilitySpec PoC (#37)
hmrc/sbt-accessibility-linter: README.md

1. Copy the template spec below into your project

```scala
import config.AppConfig
import controllers.{ContactForm, ContactHmrcForm}
import model.{AccessibilityForm, FeedbackForm, ReportProblemForm, SurveyForm}
import org.scalacheck.Arbitrary
import org.scalatest.Reporter
import org.scalatest.events.{Event, SuiteCompleted, TestPending}
import play.api.data.Form
import play.api.data.Forms._
import play.twirl.api.Html
import uk.gov.hmrc.scalatestaccessibilitylinter.views.AutomaticAccessibilitySpec
import views.html._
import scala.collection.mutable.ListBuffer

// this is the class that any consuming team would have to implement
class FrontendAccessibilitySpec
  extends AutomaticAccessibilitySpec {

  // these are things that need to have sane values for pages to render properly - there may be others
  /** example
      val appConfig: AppConfig = app.injector.instanceOf[AppConfig]
      implicit val arbConfig: Arbitrary[AppConfig] = fixed(appConfig)
  */

  // we can currently generate Arbitrary[T] but not Arbitrary[T[_]], so service teams need to provide these explicitly.
  // however, these should generally be available in the existing service codebase.

>   // these are implicit to simplify calls to render() below

Reviewer: ```suggestion
  // Note - these values are declared as `implicit` to simplify calls to `render()` below
```

  // e.g implicit val arbReportProblemPage: Arbitrary[Form[ReportProblemForm]] = fixed(reportProblemForm)

  // this is the package where the views live in the service, you may need to update this path
  // as your project might have views in a different location
  val viewPackageName = "views.html"

>   // this is the layout class which is injected into all full pages in the service
>   // and may be different in your service

Reviewer: ```suggestion
  // This is the layout class or classes which are injected into all full pages in your service.
  // This might be `HmrcLayout` or some custom class(es) that your service uses as base page templates.
```

)
```

## Plugin Setup

There are two different approaches to writing accessibility tests, you may choose the approach that

> suites your team best.

Author: Not sure if what iv put in there is good enough, might need to add some more examples but I don't know what.

> suites your team best.

Reviewer 2: @kyle-bowden on this one should it be "suits" still rather than "suites"

## Automatic accessibility testing 

Automatic accessibility testing allows you to write tests via a single spec, it can do this by

> automatically scanning through your project for views.

Reviewer 2: rather than views, what about "page templates"? I feel like we're looking inside the views package for a specific type of view, a page template

> ### Example Accessibility Spec

Reviewer 2: I wonder if we could have an outline of the procedure at the start to summarise the process, or turn it into numbered steps and with the heading "How to setup automatic accessibility testing"
1. Copy the template spec below into your project
2. Run your tests with `sbt clean a11y:test`
3. Your test output will describe what you need to add to your spec to enable testing the page templates it has been able to automatically discover in your project

Reviewer: yeah, i like this - and we need `a11y:test` to setup node etc. first time

Reviewer: do we also want to highlight that by leaving tests pending, teams can progressively cover and fix all the pages in their service over time - ie. it doesn't have to all be done as a big bang.

  extends AutomaticAccessibilitySpec {

  // these are things that need to have sane values for pages to render properly - there may be others
  /** example
      val appConfig: AppConfig = app.injector.instanceOf[AppConfig]
      implicit val arbConfig: Arbitrary[AppConfig] = fixed(appConfig)
  */

  // we can currently generate Arbitrary[T] but not Arbitrary[T[_]], so service teams need to provide these explicitly.
  // however, these should generally be available in the existing service codebase.
  // these are implicit to simplify calls to render() below
  // e.g implicit val arbReportProblemPage: Arbitrary[Form[ReportProblemForm]] = fixed(reportProblemForm)

  // this is the package where the views live in the service, you may need to update this path
  // as your project might have views in a different location
  val viewPackageName = "views.html"

  // this is the layout class which is injected into all full pages in the service
  // and may be different in your service
  val layoutClasses = Seq(classOf[views.html.components.Layout])

  // this partial function wires up the generic render() functions with arbitrary instances of the correct types.

>   // IntelliJ invariably shows some or all of these as red :(

Reviewer 2: what about, "Important: there's a known issue with intellij incorrectly displaying warnings here, you should be able to ignore these"

  /** example
      override def renderViewByClass: PartialFunction[Any, Html] = {
         case reportProblemPage: ReportProblemPage => render(reportProblemPage)
      }
  */
  
  runAccessibilityTests()
}

```

You may copy the example spec above and update it as required, to begin with you do not need to
override the `renderViewByClass` function. You can first run the spec with
```text

> sbt clean test

Reviewer 2: should this be `a11y:test`?

Author: It can be run with just `sbt test` now.

)
```

## Plugin Setup

There are two different approaches to writing accessibility tests, you may choose the approach that
suites your team best.

## How to setup automatic accessibility testing

Automatic accessibility testing allows you to write tests via a single spec, it can do this by
automatically scanning through your projects views for page templates.

### High level of how it works
* using reflection to discover page templates
* using the Play app injector to get an instance of each template
* using scalacheck/magnolia to derive arbitrary values for any template parameters
* using these arbitrary values to instantiate each page and test it for accessibility

### Which testing approach is best for your team
#### Manual accessibility testing
* Teams might already have the linter and some coverage via their view unit tests, and may want to continue down that route.
#### Automatic accessibility testing
* Teams adopting the linter for the first time might find it easier to go with the automated route.

1. Copy the template spec below into your project

> ```scala

Reviewer 2: @kyle-bowden I think if you indent the content for each step it should line it up, so:

```
1. step one

instructions

3. step two
```
becomes

1. step one

instructions

2. step two

where as if you indent it with 3 spaces like this
```
1. step one

   instructions

3. step two
```
it becomes:
1. step one

   instructions

3. step two

)
```

## Plugin Setup

There are two different approaches to writing accessibility tests, you may choose the approach that
suites your team best.

## How to setup automatic accessibility testing

Automatic accessibility testing allows you to write tests via a single spec, it can do this by
automatically scanning through your projects views for page templates.

### High level of how it works
* using reflection to discover page templates
* using the Play app injector to get an instance of each template
* using scalacheck/magnolia to derive arbitrary values for any template parameters
* using these arbitrary values to instantiate each page and test it for accessibility

### Which testing approach is best for your team
#### Manual accessibility testing
* Teams might already have the linter and some coverage via their view unit tests, and may want to continue down that route.
#### Automatic accessibility testing
* Teams adopting the linter for the first time might find it easier to go with the automated route.

1. Copy the template spec below into your project

```scala

> import config.AppConfig
> import controllers.{ContactForm, ContactHmrcForm}
> import model.{AccessibilityForm, FeedbackForm, ReportProblemForm, SurveyForm}
> import org.scalacheck.Arbitrary
> import org.scalatest.Reporter
> import org.scalatest.events.{Event, SuiteCompleted, TestPending}
> import play.api.data.Form
> import play.api.data.Forms._
> import play.twirl.api.Html
> import uk.gov.hmrc.scalatestaccessibilitylinter.views.AutomaticAccessibilitySpec
> import views.html._
> import scala.collection.mutable.ListBuffer

Reviewer: these imports need cleaning up - maybe this, to cover example code and significant things referenced in comments?
```suggestion
import org.scalacheck.Arbitrary
import play.api.data.Form
import play.twirl.api.Html
import uk.gov.hmrc.scalatestaccessibilitylinter.views.AutomaticAccessibilitySpec
import views.html._
```

> // this is the class that any consuming team would have to implement

Reviewer: this comment is probably redundant in this context?
```suggestion
```
